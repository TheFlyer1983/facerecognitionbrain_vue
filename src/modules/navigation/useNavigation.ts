import {
  RouteLocationRaw,
  useRouter
} from 'vue-router';

function useNavigation() {
  const router = useRouter();

  function navigate(route: RouteLocationRaw) {
    router.push(route);
  }

  return { navigate };
}

export default useNavigation;
