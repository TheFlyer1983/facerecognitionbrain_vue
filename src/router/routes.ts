export enum Routes {
  Home = 'home'
}