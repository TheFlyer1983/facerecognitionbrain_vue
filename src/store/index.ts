import { createStore } from 'vuex';

import { RootState } from './rootTypes';

import imageModule from './modules/image';

export const schema = {
  state: {},
  mutations: {},
  actions: {},
  modules: {
    image: imageModule
  }
};

export default createStore<RootState>(schema);
