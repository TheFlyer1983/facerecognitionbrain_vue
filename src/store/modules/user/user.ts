import { UserState, UserActionContext } from './userTypes';

import { endpoints } from '@/constants';
import { LoginInfo, RegisterInfo } from '@/types';
import {
  getAuthTokenInSession,
  saveAuthTokenInSession
} from '@/functions/storageFunctions';

export const state: UserState = {
  isSignedIn: false,
  isProfileOpen: false,
  user: null,
  token: ''
};

export const getters = {
  getUser(state: UserState): UserState['user'] {
    return state.user;
  },
  getToken(state: UserState): UserState['token'] {
    return state.token;
  },
  getIsSignedIn(state: UserState): UserState['isSignedIn'] {
    return state.isSignedIn;
  }
};

export const mutations = {
  setUserId(state: UserState, payload: string): void {
    state.user = { ...state.user, id: payload } as UserState['user'];
  },

  setUser(state: UserState, payload: UserState['user']): void {
    state.user = payload;
  },

  toggleSignIn(state: UserState, payload: UserState['isSignedIn']): void {
    state.isSignedIn = payload;
  },

  setToken(state: UserState, payload: UserState['token']): void {
    state.token = payload;
  }
};

export const actions = {
  async login(
    { commit, dispatch }: UserActionContext,
    payload: LoginInfo
  ): Promise<boolean> {
    try {
      const response = await fetch(endpoints.signin, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });

      const result = await response.json();

      if (!response.ok) throw new Error(result as string);

      saveAuthTokenInSession(result.token as string);
      commit('setToken', result.token);
      commit('setUserId', result.userId);

      const success = await dispatch('getUser', result.userId as string);
      if (!success) throw new Error('Error');

      return true;
    } catch (error) {
      console.log(error, 'In the error block');
      return false;
    }
  },
  async getUser({ commit }: UserActionContext, payload: string) {
    const requestURL = endpoints.getProfile.replace(':id', payload);
    const token = getAuthTokenInSession();
    try {
      if (!token) throw new Error('No valid token');
      const response = await fetch(requestURL, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          Authorization: token
        }
      });
      const result = await response.json();
      if (!response.ok) throw new Error(result);

      commit('setUser', result);
      commit('toggleSignIn', true);
      return true;
    } catch (error) {
      console.log(error);
      return false;
    }
  },
  async registerUser(
    { commit }: UserActionContext,
    payload: RegisterInfo
  ): Promise<boolean> {
    try {
      const response = await fetch(endpoints.register, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });

      const result = await response.json();

      if (!response.ok) throw new Error(result as string);

      saveAuthTokenInSession(result.session.token as string);
      commit('setUser', result.register.user);

      return true;
    } catch (error) {
      console.log(error);
      return false;
    }
  },
  getToken({ commit, dispatch }: UserActionContext) {
    const token = getAuthTokenInSession();

    if (token) {
      commit('setToken', token);
      dispatch('authenticated');
    }
  },
  async authenticated({ state, dispatch }: UserActionContext) {
    const token = state.token;

    if (token) {
      try {
        const response = await fetch(endpoints.signin, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            Authorization: token
          }
        });

        const result = await response.json();

        if (!response.ok) throw new Error(result as string);

        dispatch('getUser', result.id);
      } catch (error) {
        console.log(error);
      }
    }
  }
};

const userModule = {
  namespaced: true,
  state,
  getters,
  mutations,
  actions
};

export default userModule;
