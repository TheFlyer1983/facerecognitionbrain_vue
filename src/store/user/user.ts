import { defineStore } from 'pinia';
import { UpdateInfo, User, UserState } from './userTypes';
import {
  LoginInfo,
  LoginResponse,
  RegisterInfo,
  RegisterResponse
} from '@/types';
import request from '@/functions/request';
import { endpoints } from '@/constants';
import {
  getAuthTokenInSession,
  removeAuthTokenFromSession,
  saveAuthTokenInSession
} from '@/functions/storageFunctions';

export const useUserStore = defineStore('user', {
  state: (): UserState => ({
    isSignedIn: false,
    isProfileOpen: false,
    id: '',
    user: null,
    token: '',
    rank: ''
  }),

  actions: {
    async login(payload: LoginInfo) {
      payload = { ...payload, returnSecureToken: true };

      try {
        const response = await request.post<LoginResponse>(
          endpoints.signin,
          payload,
          {
            params: { key: import.meta.env.VITE_APP_FIREBASE_API_KEY }
          }
        );

        saveAuthTokenInSession(response.data.idToken);
        this.token = response.data.idToken;
        this.id = response.data.localId;

        const success = await this.getUser(response.data.localId);
        if (!success) throw new Error('Error');

        return true;
      } catch (error) {
        console.error(error);
        return false;
      }
    },

    async getUser(userId: string) {
      const requestURL = endpoints.profile.replace(':id', userId);

      try {
        const response = await request.get<User>(requestURL, {
          params: { auth: this.token }
        });

        this.user = response.data;
        this.isSignedIn = true;

        //this.getRank();

        return true;
      } catch (error) {
        console.error(error);
        return false;
      }
    },

    async registerUser(payload: RegisterInfo) {
      const registerPayload = {
        email: payload.email,
        password: payload.password,
        returnSecureToken: true
      };
      try {
        const response = await request.post<RegisterResponse>(
          endpoints.register,
          registerPayload,
          {
            params: { key: import.meta.env.VITE_APP_FIREBASE_API_KEY }
          }
        );

        saveAuthTokenInSession(response.data.idToken);
        this.token = response.data.idToken;
        this.id = response.data.localId;
        await this.createProfile(payload.name);

        return true;
      } catch (error) {
        console.error(error);
        return false;
      }
    },

    async createProfile(name: string) {
      const profilePayload = {
        name,
        entries: 0,
        joined: new Date()
      };

      const requestURL = endpoints.profile.replace(':id', this.id);
      try {
        const response = await request.put(requestURL, profilePayload, {
          params: { auth: this.token }
        });

        this.user = response.data;
      } catch (error) {
        console.log(error);
      }
    },

    getToken() {
      const token = getAuthTokenInSession();

      if (token) {
        this.token = token;

        this.authenticated();
      }
    },

    async authenticated() {
      if (this.token) {
        try {
          const response = await request.post(endpoints.signin);

          this.getUser(response.data.id);
        } catch (error) {
          console.error(error);
        }
      }
    },

    async getRank() {
      const entries = this.user?.entries;

      try {
        const response = await request.post(endpoints.rank, { entries });
        this.rank = response.data;
      } catch (error) {
        console.error(error);
      }
    },

    signout() {
      removeAuthTokenFromSession();
      this.$reset();
    },

    async updateUser(payload: UpdateInfo) {
      const requestURL = endpoints.profile.replace(':id', this.id);
      try {
        await request.put(requestURL, payload);

        this.getUser(this.id as string);
      } catch (error) {
        console.error(error);
      }
    },

    async deleteUser() {
      const requestURL = endpoints.profile.replace(':id', this.id);
      try {
        await request.delete(requestURL, { params: { auth: this.token } });

        await request.post(
          endpoints.delete,
          { idToken: this.id },
          { params: { key: import.meta.env.VITE_APP_FIREBASE_API_KEY } }
        );

        this.signout();
      } catch (error) {
        console.error(error);
      }
    }
  }
});
