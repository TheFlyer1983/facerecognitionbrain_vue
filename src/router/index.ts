import { createRouter, createWebHistory, RouteRecordRaw } from 'vue-router';

import HelloWorld from '@/views/HelloWorld.vue';
import { Routes } from './routes';

export const INITIAL_ROUTE = '/';

const routes: Array<RouteRecordRaw> = [
  {
    path: INITIAL_ROUTE,
    name: 'Index',
    redirect: () => ({ name: Routes.Home })
  },
  {
    path: '/home',
    name: Routes.Home,
    component: HelloWorld,
    props: { msg: 'Hello Vue 3 + TypeScript + Vite' }
  },
  {
    path: '/login',
    name: Routes.Login,
    component: () => import('@/views/Login/Login.vue')
  },
  {
    path: '/register',
    name: Routes.Register,
    component: () => import('@/views/Register/Register.vue')
  }
];

const router = createRouter({
  history: createWebHistory(INITIAL_ROUTE),
  routes
});

export default router;
