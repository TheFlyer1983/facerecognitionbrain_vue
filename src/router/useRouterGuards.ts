import { getAuthTokenInSession } from '@/functions/storageFunctions';
import { NavigationGuardNext } from 'vue-router';
import { Routes } from './routes';

function useRouterGuards() {
  function authenticated(next: NavigationGuardNext) {
    const { token } = getAuthTokenInSession();

    return token ? next() : next({ name: Routes.Login });
  }

  return { authenticated };
}

export default useRouterGuards;
