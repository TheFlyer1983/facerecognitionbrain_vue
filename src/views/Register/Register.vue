<script setup lang="ts">
import { ref } from 'vue';
import { useStore } from 'vuex';
import { useNavigation } from '@/modules/navigation';
import { Routes } from '@/router/routes';

const store = useStore();
const { navigate } = useNavigation();

const name = ref('');
const email = ref('');
const password = ref('');

async function onSubmitRegister(): Promise<void> {
  const registerInfo = {
    name: name.value,
    email: email.value,
    password: password.value
  }
  const success = await store.dispatch('user/registerUser', registerInfo);
  if (success) {
    navigate({ name: Routes.Home });
  }
}
</script>

<template>
  <article class="register-article">
    <main class="register-article-main">
      <div class="register-article-main-measure">
        <fieldset class="register-article-main-measure-fieldset">
          <legend class="register-article-main-measure-fieldset--legend">
            Register
          </legend>
          <div class="register-article-main-measure-fieldset__field">
            <label
              for="name"
              class="register-article-main-measure-fieldset__field--label"
            >
              Name
            </label>
            <input
              type="text"
              name="name"
              id="name"
              v-model="name"
              class="register-article-main-measure-fieldset__field--input"
            />
          </div>
          <div lass="register-article-main-measure-fieldset__field">
            <label
              for="email-address"
              class="register-article-main-measure-fieldset__field--label"
            >
              Email
            </label>
            <input
              type="email"
              name="email-address"
              id="email-address"
              v-model="email"
              class="register-article-main-measure-fieldset__field--input"
            />
          </div>
          <div lass="register-article-main-measure-fieldset__field">
            <label
              for="password"
              class="register-article-main-measure-fieldset__field--label"
            >
              Password
            </label>
            <input
              type="password"
              name="password"
              id="password"
              v-model="password"
              class="register-article-main-measure-fieldset__field--input"
            />
          </div>
        </fieldset>
        <div>
          <input
            type="submit"
            value="Register"
            @click="onSubmitRegister"
            class="register-article-main-measure--submit"
          />
        </div>
      </div>
    </main>
  </article>
</template>

<style scoped lang="scss">
.register-article {
  border-radius: 0.5rem;
  border: solid 1px var(--black-10);
  width: 100%;
  max-width: 32rem;
  box-shadow: 4px 4px 8px 0 rgba(0, 0, 0, 0.2);
  margin: 4rem auto;

  @include breakpoint-medium {
    width: 50%;
  }

  @include breakpoint-large {
    width: 25%;
  }

  &-main {
    padding: 2rem;
    color: var(--black-80);

    &-measure {
      max-width: 30em;

      &-fieldset {
        border: solid 1px transparent;
        padding-left: 0;
        padding-right: 0;
        margin-left: 0;
        margin-right: 0;

        &--legend {
          font-size: 3rem;
          font-weight: 600;
          padding-left: 0;
          padding-right: 0;
          margin-left: 0;
          margin-right: 0;
        }

        &__field {
          margin-top: 1rem;

          &--label {
            display: block;
            font-weight: 600;
            line-height: 1.5;
            font-size: 0.875rem;
          }

          &--input {
            -webkit-appearance: none;
            -moz-appearance: none;
            padding: 0.5rem;
            border: solid 1px;
            width: 50%;
            background-color: transparent;

            &:hover {
              background-color: black;
              color: white;
            }
          }
        }
      }

      &--submit {
        -webkit-appearance: none;
        -moz-appearance: none;
        -moz-osx-font-smoothing: grayscale;
        cursor: pointer;
        background-color: transparent;
        border: solid 1px black;
        font-size: 0.875rem;
        display: inline-block;
        backface-visibility: hidden;
        transform: translateZ(0);
        transition: transform 0.25 ease-out;
        padding-left: 1rem;
        padding-right: 1rem;

        &:hover {
          transform: scale(1.05);
        }
      }

      &__register {
        line-height: 1.5;
        margin-top: 1rem;

        &--link {
          cursor: pointer;
          display: block;
          opacity: 1;
          transition: opacity 0.15s ease-in;
          font-size: 0.875rem;
          color: black;
          text-decoration: none;

          &:hover {
            opacity: 0.5;
            transition: opacity 0.15s ease-in;
          }
        }
      }
    }
  }
}
</style>
