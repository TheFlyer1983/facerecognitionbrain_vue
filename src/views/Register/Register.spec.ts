import { describe, expect, it, beforeEach, vi, beforeAll } from 'vitest';
import { shallowMount, VueWrapper } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import { useUserStore } from '@/store/user';
import { useNavigation } from '@/modules/navigation';
import { Routes } from '@/router/routes';

import Register from './Register.vue';
import { UserMock } from '@/fixtures/users';

vi.mock('@/modules/navigation');

const mockedUseNavigation = vi.mocked(useNavigation, true);

const navigateMock = vi.fn();

const mockedRegisterInfo = {
  name: 'Paul',
  email: 'test@test.com',
  password: 'password'
};

const pinia = createTestingPinia();
const mockUserStore = useUserStore(pinia);

describe('Given the `Register` component', () => {
  const render = () => shallowMount(Register, { global: { plugins: [pinia] } });
  let wrapper: VueWrapper<InstanceType<typeof Register>>;

  beforeAll(() => {
    mockedUseNavigation.mockReturnValue({
      navigate: navigateMock
    });
  });

  describe('when it is rendered', () => {
    beforeEach(() => {
      wrapper = render();
    });

    it('should match the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    describe('when the submit button is clicked', () => {
      beforeEach(() => {
        wrapper.vm.name = mockedRegisterInfo.name;
        wrapper.vm.email = mockedRegisterInfo.email;
        wrapper.vm.password = mockedRegisterInfo.password;

        mockUserStore.$patch({ user: { ...UserMock } });

        wrapper.find('[data-test="submit"]').trigger('click');
      });

      it('should dispatch the correct action', () => {
        expect(mockUserStore.registerUser).toHaveBeenCalledWith(
          mockedRegisterInfo
        );
      });

      it('should navigate to the correct page', () => {
        expect(navigateMock).toHaveBeenCalledWith({ name: Routes.Home });
      });
    });

    describe.each([
      {
        field: 'name',
        value: 'Paul',
        id: 'name'
      },
      {
        field: 'email',
        value: 'test@test.com',
        id: 'email-address'
      },
      {
        field: 'password',
        value: 'password',
        id: 'password'
      }
    ])('when the inputs are updated', (arg) => {
      describe(`when the ${arg.field} is updated`, () => {
        beforeEach(() => {
          wrapper.find(`[id="${arg.id}"]`).setValue(arg.value);
        });

        it(`should update the ${arg.field} property`, () => {
          expect(wrapper.vm[arg.field]).toBe(arg.value);
        });
      });
    });
  });
});
