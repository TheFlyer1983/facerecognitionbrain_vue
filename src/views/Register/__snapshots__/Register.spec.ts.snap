// Vitest Snapshot v1

exports[`Given the \`Register\` component > when it is rendered > should match the snapshot 1`] = `
<article
  class="register-article"
  data-v-89291785=""
>
  <main
    class="register-article-main"
    data-v-89291785=""
  >
    <div
      class="register-article-main-measure"
      data-v-89291785=""
    >
      <fieldset
        class="register-article-main-measure-fieldset"
        data-v-89291785=""
      >
        <legend
          class="register-article-main-measure-fieldset--legend"
          data-v-89291785=""
        >
           Register 
        </legend>
        <div
          class="register-article-main-measure-fieldset__field"
          data-v-89291785=""
        >
          <label
            class="register-article-main-measure-fieldset__field--label"
            data-v-89291785=""
            for="name"
          >
             Name 
          </label>
          <input
            class="register-article-main-measure-fieldset__field--input"
            data-v-89291785=""
            id="name"
            name="name"
            type="text"
          />
        </div>
        <div
          data-v-89291785=""
          lass="register-article-main-measure-fieldset__field"
        >
          <label
            class="register-article-main-measure-fieldset__field--label"
            data-v-89291785=""
            for="email-address"
          >
             Email 
          </label>
          <input
            class="register-article-main-measure-fieldset__field--input"
            data-v-89291785=""
            id="email-address"
            name="email-address"
            type="email"
          />
        </div>
        <div
          data-v-89291785=""
          lass="register-article-main-measure-fieldset__field"
        >
          <label
            class="register-article-main-measure-fieldset__field--label"
            data-v-89291785=""
            for="password"
          >
             Password 
          </label>
          <input
            class="register-article-main-measure-fieldset__field--input"
            data-v-89291785=""
            id="password"
            name="password"
            type="password"
          />
        </div>
      </fieldset>
      <div
        data-v-89291785=""
      >
        <input
          class="register-article-main-measure--submit"
          data-test="submit"
          data-v-89291785=""
          type="submit"
          value="Register"
        />
      </div>
    </div>
  </main>
</article>
`;
